hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,531 @@
+	if err != nil {
+		return ReconcileActionNoAction, NoAction, errors.Wrap(err, "comparing resource hash")
+	}
+
+	if !hasChanged && data.IsTerminalProvisioningState() {
+		msg := fmt.Sprintf("resource spec has not changed and resource is in terminal state: %q", state)
+		data.log.V(1).Info(msg)
+		return ReconcileActionNoAction, NoAction, nil
+	}
+
+	if state == armclient.DeletingProvisioningState {
+		return ReconcileActionNoAction, NoAction, errors.Errorf("resource is currently deleting; it can not be applied")
+	}
+
+	if data.GetDeploymentIdOrDefault() != "" {
+		// There is an ongoing deployment we need to monitor
+		return ReconcileActionWatchDeployment, gr.MonitorDeployment, nil
+	}
+
+	return ReconcileActionBeginDeployment, gr.CreateDeployment, nil
+}
+
+//////////////////////////////////////////
+// Actions
+//////////////////////////////////////////
+
+func NoAction(ctx context.Context, action ReconcileAction, data *ReconcileMetadata) (ctrl.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+// StartDeleteOfResource will begin the delete of a resource by telling Azure to start deleting it. The resource will be
+// marked with the provisioning state of "Deleting".
+func (gr *GenericReconciler) StartDeleteOfResource(
+	ctx context.Context,
+	action ReconcileAction,
+	data *ReconcileMetadata) (ctrl.Result, error) {
+
+	msg := "Starting delete of resource"
+	data.log.Info(msg)
+	gr.Recorder.Event(data.metaObj, v1.EventTypeNormal, string(action), msg)
+
+	resource, err := gr.constructArmResource(ctx, data)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "converting %q to armResourceSpec", data.metaObj.GetName())
+	}
+
+	err = gr.Patch(ctx, data, func(ctx context.Context, mutData *ReconcileMetadata) error {
+		if resource.GetId() != "" {
+			emptyStatus, err := NewEmptyArmResourceStatus(mutData.metaObj)
+			if err != nil {
+				return errors.Wrapf(err, "creating empty status for %q", resource.GetId())
+			}
+
+			err = gr.ARMClient.BeginDeleteResource(ctx, resource.GetId(), resource.Spec().GetApiVersion(), emptyStatus)
+			if err != nil {
+				return errors.Wrapf(err, "deleting resource %q", resource.Spec().GetType())
+			}
+			data.SetResourceProvisioningState(armclient.DeletingProvisioningState)
+		} else {
+			controllerutil.RemoveFinalizer(mutData.metaObj, GenericControllerFinalizer)
+		}
+
+		return nil
+	})
+
+	err = client.IgnoreNotFound(err)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "patching after delete")
+	}
+
+	// delete has started, check back to seen when the finalizer can be removed
+	return ctrl.Result{
+		RequeueAfter: 5 * time.Second,
+	}, nil
+}
+
+// MonitorDelete will call Azure to check if the resource still exists. If so, it will requeue, else,
+// the finalizer will be removed.
+func (gr *GenericReconciler) MonitorDelete(
+	ctx context.Context,
+	action ReconcileAction,
+	data *ReconcileMetadata) (ctrl.Result, error) {
+
+	msg := "Continue monitoring ongoing delete"
```

Comment:
Not entirely sure of the wording, but this log message should be improved. #grammapedant
```suggestion
	msg := "Continue monitoring of deletion"
```